Updating workflow json templates

amaltaro opened the pull request:
Fixes #11860 

#### Status
In development

#### Description
This PR provides a full refresh of all our workflow test json templates. Each commit should be refreshing a different workflow spec.

In addition, I had to make a minor change to the `validate-test-wfs.py` script, otherwise the whole execution was terminating upon a "broken" DQMGui response.

Lastly, I have also updated the list of pileups used by these templates, which we need to register into our Dev instances of MSPileup (testbed is already up-to-date).

Breakdown of the templates to be created are described in:
https://github.com/dmwm/WMCore/issues/11860

ToDo list:
- [x] ReReco workflows
- [x] DQMHarvest workflows
- [x] TaskChain workflows
- [x] StepChain workflows
- [x] Figure out how to gracefully reuse the AcqEra for ReReco_LumiMask.json
- [ ] If available in time, update one of the CMSSW releases to pre4 in the following templates: MCRecyc_PU.json and MCRecyc.json
- [x] Lock input and parent data for DMWM/ReReco_RunBlocklists_Parents both at CERN and FNAL

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
None

#### External dependencies / deployment changes
Summary of the templates is as follows:
**ReReco**
- Template 1: with block and run whitelist, include parents and multicore, 
- Template 2: with skims, with LumiList, with automatic harvesting
- Template 3: requiring GPUs, with unprocessable blocks
NOTE: a workflow processing the full dataset would probably be too large.

**DQMHarvest**
- Template 1: harvesting a full data DQMIO dataset
- Template 2: with run whitelist, set to harvest in byRun mode
- Template 3: with LumiList, set to harvest in multirun mode

**TaskChain**
- Template 1: MC reading LHE files (PFN remote read)
- Template 2: MC from scratch, MC extension, MC automatic harvesting, with transient output, output rules created by WMAgent
- Template 3: MC from scratch, with Pileup, with duplicate output module, different Multicore/Memory/EventStreams/MaxPSS
- Template 4: MC recycling, PrimaryDataset override, different AcqEra/ProcStr/ProcVer/PrepID/CMSSW
- Template 5: MC recycling, with Pileup, with KeepOutput false
- Template 6: Data workflow, with !IncludeParents,  with LumiList input --> Data automatic harvesting missing!!!
- Template 7: MC recycling MINIAODSIM - hence with Dataset start policy
- Template 8: MC from scratch, with Keep output, with GPU requirements
- Bonus: added a template that uses SONIC inference infrastructure

**StepChain (almost a dup of the TaskChain templates - or vice-versa)**
- Template 1: MC reading LHE files
- Template 2: MC from scratch, MC extension, with transient output
- Template 3: MC from scratch, with Pileup, MC automatic harvesting, different Multicore/Memory/EventStreams/MaxPSS
- Template 4: MC recycling, PrimaryDataset override, different AcqEra/ProcStr/ProcVer/PrepID/CMSSW
- Template 5: MC recycling, with Pileup, with KeepOutput false
- Template 6: Data workflow, with !IncludeParents,  with LumiList input, Data automatic harvesting
- Template 7: MC recycling MINIAODSIM - hence with Dataset start policy
- Others: with duplicate output module, with multiple PU datasets
NOTE: I think PrimaryDataset override is not supported - maybe ProcVersion for different steps is also not supported

amaltaro committed:
Refactor all ReReco templates

Two fully-featured ReReco templates; remove unneded; move others to Integration

Remove unneeded ReReco template

Remove unneeded parents ReReco workflow

Enable TrustSitelists for a Rereco parent template

Fix comments for ReReco template

Update rereco template

Refining ReReco templates

amaltaro committed:
Refactor all StepChain templates

Set AAA for secondary

Update StepChain templates

Copied converted TaskChains to StepChain templates

Fix unsupported keys in StepChain templates

Refining SC templates

Fix memory for stepchain workflows

amaltaro committed:
Update list of dev pileups in MSPileup

Add new pileup to pileup_dev.json file

amaltaro committed:
Refresh all DQMHarvest templates

update number of lumis

remove a closing brackets

Co-authored-by: Dario Mapelli <mapelli.dario@gmail.com>

Refining DQMHarvest workflows

amaltaro commented:
I have migrated a script from my repository over to `wmcore-tools`:
https://gitlab.cern.ch/dmwm/wmcore-tools/-/blob/master/mstransferor/makeRucioRules.py?ref_type=heads

and used it to lock the DQMIO dataset both at FNAL and CERN, e.g.:
```
(WMAgent-2.3.7.6) [xxx@vocmsxxx:logs]$ python makeRucioRules.py -c /JetMET0/Run2024D-2024CDEReprocessing-v1/DQMIO -r T1_US_FNAL_Disk
2025-01-30 21:00:09,584:INFO:Rucio: WMCore Rucio initialization parameters: ...
2025-01-30 21:00:09,585:INFO:Rucio: Rucio client initialization parameters: ...
2025-01-30 21:00:09,585:INFO:makeRucioRules: 
Creating rule for DID: /JetMET0/Run2024D-2024CDEReprocessing-v1/DQMIO, with RSE: T1_US_FNAL_Disk and other attrs: {'copies': 1, 'activity': 'Production Input', 'lifetime': None, 'account': 'wmcore_transferor', 'grouping': 'ALL', 'comment': 'WMCore MSTransferor input data placement'}
2025-01-30 21:00:09,647:INFO:makeRucioRules: Response: ['5d3b05b9a8e5419a814e2ef0f1bf7056']
```

FYI @mapellidario

mapellidario commented on test/data/ReqMgr/requests/Integration/DQMHarvesting_LumiMask_MultiRun.json:
```diff
@@ -0,0 +1,49 @@
+        "DbsUrl": "https://cmsweb.cern.ch/dbs/prod/global/DBSReader", 
+        "GlobalTag": "140X_dataRun3_v20",
+        "InputDataset": "/JetMET0/Run2024D-2024CDEReprocessing-v1/DQMIO",
+        "LumiList": {"380306": [[28, 35], [41, 50], [60, 60]], "380307": [[5, 10], [20, 21]], "380346": [[61, 70]]]},
```
```suggestion
        "LumiList": {"380306": [[28, 35], [41, 50], [60, 60]], "380307": [[5, 10], [20, 21]], "380346": [[61, 70]]},
```

we need to remove a closing square bracket

amaltaro replied:
Thank you, Dario. Those problems always sneak through!

amaltaro commented:
Rules for the ReReco templates have been created:
```
Creating rule for DID: ['/MinimumBias/Run2023C-22Sep2023_v4-v1/MINIAOD#146eb70b-12ad-47dd-a109-b7feef4927cf', '/MinimumBias/Run2023C-22Sep2023_v4-v1/MINIAOD#5b047ae4-f01e-44b4-b960-ae443ed46c2e', '/MinimumBias/Run2023C-PromptReco-v4/AOD#8ba900bd-a353-47d6-aeca-e81e875e2690', '/MinimumBias/Run2023C-PromptReco-v4/AOD#859fe3fb-7402-49a2-9345-8f37b09de617', '/MinimumBias/Run2023C-PromptReco-v4/AOD#0880f752-9eb5-4d4c-88cc-e1c49d2d7ebe', '/MinimumBias/Run2023C-PromptReco-v4/AOD#ae818ccd-2395-43a8-be51-5d87f8c175b3'], with RSE: T1_US_FNAL_Disk and other attrs: {'copies': 1, 'activity': 'Production Input', 'lifetime': None, 'account': 'wmcore_transferor', 'grouping': 'DATASET', 'comment': 'WMCore MSTransferor input data placement'}
2025-02-01 12:31:45,577:INFO:makeRucioRules: Response: ['496efe034ffc487b8e664c0566d95d6e', '84ed8be94a0a4e989e173473684b3bfe', 'ce775b70dc094023a15e25d3a9a2d36f', 'c581cad573794a05aaec3030d9665174', '0576ca4bb7414a69b5d2c761d0cffc6e', '9b16d5edbd8649e19af65e44f91f1afb']
```

and
```
Creating rule for DID: /MuonEG/Run2022C-v1/RAW#00353169-1291-4f60-a332-b8d66d2e1647, with RSE: T2_CH_CERN and other attrs: {'copies': 1, 'activity': 'Production Input', 'lifetime': None, 'account': 'wmcore_transferor', 'grouping': 'DATASET', 'comment': 'WMCore MSTransferor input data placement'}
2025-02-01 12:55:11,708:INFO:makeRucioRules: Response: ['c833adf9da8646daa2cdb5d44666897a']
```

amaltaro committed:
Refactor all TaskChain templates

Correct TaskChain unit tests, moving a template over to Static dir

Fix AcqEra with dashes

Change InputDataset and ConfigCacheId for TC_MCRecyc templates

Change TC_MCRecyc templates to use DBS Prod for data discovery/location

Fix campaign for TC_Parent_LumiMask

Fix DQM url for TC_ProdPsi

Fix comments and a couple of settings for TC

Update taskchain templates

GPU workflows for Purdue/SONIC tests

Refining TaskChain templates

Fix memory for TaskChain workflows

amaltaro committed:
Safe dict access for validate-test-wfs script

amaltaro commented:
@mapellidario I asked for your review, but please spare yourself of looking into the gazillions templates. There is one or two actual code changes together with this PR. Please let me know if you have any questions/concerns.

amaltaro committed:
Reuse ReReco AcqEra during assignment
